File: windows/GitProxy.cs
using System.Diagnostics;

namespace SpeedLane;

/// <summary>为 git 命令行按域名设置代理(只影响列表内域名,其余仓库不走代理)</summary>
public static class GitProxy
{
    public static void Apply(AppSettings settings, string proxyUrl)
    {
        Clear(settings);
        var applied = new List<string>();
        foreach (var domain in settings.ActiveDomains)
        {
            if (RunGit("config", "--global", $"http.https://{domain}/.proxy", proxyUrl))
                applied.Add(domain);
        }
        settings.AppliedGitDomains = applied;
        settings.Save();
    }

    public static void Clear(AppSettings settings)
    {
        foreach (var domain in settings.AppliedGitDomains)
            RunGit("config", "--global", "--unset-all", $"http.https://{domain}/.proxy");
        settings.AppliedGitDomains = new List<string>();
        settings.Save();
    }

    private static bool RunGit(params string[] args)
    {
        try
        {
            var psi = new ProcessStartInfo
            {
                FileName = "git",
                UseShellExecute = false,
                CreateNoWindow = true,
                RedirectStandardOutput = true,
                RedirectStandardError = true,
            };
            foreach (var a in args) psi.ArgumentList.Add(a);
            using var process = Process.Start(psi)!;
            process.WaitForExit(10000);
            return process.HasExited && process.ExitCode == 0;
        }
        catch
        {
            // 未安装 git 时静默跳过
            return false;
        }
    }
}

File: Sources/SpeedLane/GitProxy.swift
import Foundation

/// 为 git 命令行按域名设置代理(只影响列表内的域名,其余仓库不走代理)
/// 例如:git config --global http.https://github.com/.proxy socks5h://127.0.0.1:1080
enum GitProxy {
    private static let git = "/usr/bin/git"
    private static let appliedKey = "appliedGitDomains"

    static func apply(domains: [String], proxyURL: String) {
        clear()
        var applied: [String] = []
        for domain in domains {
            let (status, _) = runCommand(git, [
                "config", "--global",
                "http.https://\(domain)/.proxy", proxyURL,
            ])
            if status == 0 { applied.append(domain) }
        }
        UserDefaults.standard.set(applied, forKey: appliedKey)
    }

    static func clear() {
        let applied = UserDefaults.standard.stringArray(forKey: appliedKey) ?? []
        for domain in applied {
            runCommand(git, [
                "config", "--global",
                "--unset-all", "http.https://\(domain)/.proxy",
            ])
        }
        UserDefaults.standard.removeObject(forKey: appliedKey)
    }
}
